core/src/main/java/org/radargun/stages/test/legacy/LegacyStressor.java

```diff
@@ -87,18 +95,22 @@ private void runInternal() {
                if (ongoingTx != null) {
                   endTransactionAndRegisterStats(null);
                }
-               stats.begin();
-               this.started = started;
                break;
-            }
-            Operation operation = operationSelector.next(random);
-            try {
-               logic.run(operation);
-            } catch (OperationLogic.RequestException e) {
-               // the exception was already logged in makeRequest
+            } else {
+               if (!isStarted()) {
+                  threadCountDown.countDown();
+                  this.started = true;
```

Comment:
Okay, this looks much better. The only thing is that I would set `this.started = true` only when we see that `stage.isStarted()` - put that next to `stats.begin()`. For the countdown check, you could use just a local var.
Then, in `recording()` you won't have to check the `stage.started` which is a volatile read.